### ISSUE TITLE
[Bug]: "env" hash mixes upper and lower case Variables

### ISSUE THREAD

**Issue by brosmar** [jenkinsci/templating-engine-plugin#320]
### Jenkins Version

CloudBees CI Client Controller 2.387.3.5-rolling

### JTE Version

2.5.3

### Bug Description

I am not able to set all necessary proxy environment variables from a templating script.

I have to set the following environment variables:

HTTP_PROXY
HTTPS_PROXY
NO_PROXY
http_proxy
https_proxy
no_proxy

the env hash dose not clearly distinguish between lower and capital letters.
In the attached test cycle I am only able to set one of the above Variables in the system environment from the script.
Some of the current Environment variable could even not be replaced.

Please implement a proper small and capital letter handling of the system environment.

If I miss something please advise!

Thank you

### Relevant log output

```shell
[Pipeline] echo
--------------------------------------------------
[Pipeline] echo
--------------------------------------------------
[Pipeline] echo
--------------------------------------------------
[Pipeline] echo
--------------------------------------------------
[Pipeline] echo
Environment TEST:
[Pipeline] echo
--------------------------------------------------
[Pipeline] sh
+ sort
+ env
+ grep -i proxy
HTTP_PROXY11=11
HTTP_PROXY=http://squid-proxy.gcp.cloud.io:3128/
HTTPS_PROXY12=12
https_proxy=http://squid-proxy.gcp.cloud.io:3128/
NO_PROXY13=13
NO_PROXY=3
NO_PROXY_WORKAROUND=.cedelgroup.com,.deutsche-boerse.de,.dbgcloud.io,.internal,.xeop.de,.svc,.local,localhost,127.0.0.1,169.254.169.254,sts.deutsche-boerse.com,.googleapis.com,.ltsapis.goog,.gcr.io,.gstatic.com,.appspot.com,.cloudfunctions.net,.pki.goog,.cloudproxy.app,.run.app,.ci.dom,.devpnrad.net,.ondbag.com,packages.cloud.google.com,gcr.io,appengine.google.com,pki.goog
[Pipeline] echo
setting capital letter proxy settings
[Pipeline] sh
+ sort
+ grep -i proxy
+ env
HTTP_PROXY11=11
HTTP_PROXY=http://squid-proxy.gcp.dbgcloud.io:3128/
HTTPS_PROXY12=12
https_proxy=http://squid-proxy.gcp.dbgcloud.io:3128/
NO_PROXY13=13
NO_PROXY=3
[Pipeline] echo
setting small letter proxy settings
[Pipeline] echo
--------------------------------------------------
[Pipeline] echo
Environment after TEST:
[Pipeline] echo
--------------------------------------------------
[Pipeline] sh
+ env
+ grep -i proxy
+ sort
HTTP_PROXY11=11
HTTP_PROXY=http://squid-proxy.gcp.cloud.io:3128/
HTTPS_PROXY12=12
https_proxy=http://squid-proxy.gcp.cloud.io:3128/
NO_PROXY13=13
NO_PROXY=localhost,127.0.0.1
```

### Steps to Reproduce

The following script is Producing the above output

 ```shell
def testOutput() {
    println "--------------------------------------------------"
    println "--------------------------------------------------"
    println "--------------------------------------------------"
    println "--------------------------------------------------"
    println "Environment TEST:"
    println "--------------------------------------------------"
    env.http_proxy='1'
    env.https_proxy='2'
    env.no_proxy='3'
    env.HTTP_PROXY='4'
    env.HTTPS_PROXY='5'
    env.NO_PROXY='6'
    env.http_proxy11='11'
    env.https_proxy12='12'
    env.no_proxy13='13'
    env.HTTP_PROXY11='14'
    env.HTTPS_PROXY12='15'
    env.NO_PROXY13='16'

    sh 'env | grep -i proxy | sort'


    println "setting capital letter proxy settings"
    env.HTTP_PROXY='http://myproxydefault'
    env.HTTPS_PROXY='http://myproxydefault'
    env.NO_PROXY='localhost,127.0.0.1'

    sh 'env | grep -i proxy | sort'
    println "setting small letter proxy settings"
    env.http_proxy='http://myproxydefault'
    env.https_proxy='http://myproxydefault'
    env.no_proxy='localhost,127.0.0.1'

    println "--------------------------------------------------"
    println "Environment after TEST:"
    println "--------------------------------------------------"
    sh 'env | grep -i proxy | sort'
    println "--------------------------------------------------"
    println "--------------------------------------------------"
    println "--------------------------------------------------"
}

```

**Comment by brosmar** [comment]
I just realized that the Relevant Log output was false. I am now uploaded a new test and test output. I hope it will show the problem.

**Comment by steven-terrana** [comment]
hey @brosmar - after some digging, this isn't actually a bug with JTE. 

It's actually intended behavior according to Jenkins itself.

Here's the rabbit hole - 
1. when you reference the `env` var - you're actually referencing the [`EnvActionImpl`](https://github.com/jenkinsci/workflow-cps-plugin/blob/master/plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/EnvActionImpl.java) class. 
2. `EnvActionImpl` makes use of [`EnvironmentExpander`](https://github.com/jenkinsci/workflow-step-api-plugin/blob/master/src/main/java/org/jenkinsci/plugins/workflow/steps/EnvironmentExpander.java) when fetching environment variables to proxy access requests to the actual [`EnvVars`](https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/EnvVars.java) class
3. The `EnvVars` class explicitly states that environment variables should be case insensitive but case preserving ([source](https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/EnvVars.java#L62-L63))
4. This gets implemented by `EnvVars` being a [subclass of TreeMap](https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/EnvVars.java#L75C23-L75C23) and using a [case insensitive comparator](https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/EnvVars.java#L108) to lookup values. 

TL;DR -> basically, both values get stored in the TreeMap but one of the values will never be returned because of how keys are sorted.

other helpful links:
* https://www.baeldung.com/java-map-with-case-insensitive-keys 

<hr>

Closing this issue as it isn't related to JTE.

**Comment by brosmar** [comment]
If this is expected behavior than the environment can not be set properly. 

We are using software that needs both declarations. HTTPS_PROXY and https_proxy in the environment. 

So we are lost.
